From 179d3e912e7554aba5a12417182b0a60721288c0 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Fri, 11 May 2018 17:15:45 -0700
Subject: [PATCH 0355/1852] drm: xlnx: zynqmp: Don't update the plane for the
 same fb

commit 7cb966bab6642a2801c9e5f1bd222141008785ac from
https://github.com/Xilinx/linux-xlnx.git

The async plane update triggers the other plane's atomic_async_update
if the plane is enabled for the same crtc, even though there's no change
for that plane. Depending on the timing of requests, this can
result in 2 back-to-back descriptors for a single frame that will be
queued in the dma descriptor list and consumed sequentially by dmaengine.
In this case, the frame will be displayed twice, and it will end up
displaying the backbuffer.

This patch fixes it by skipping plane update for the same fb. There's
no reason to create and schedule another descriptor for the same fb.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Tested-by: Daniel Steger <dsteger@xilinx.com>
Reviewed-by: Satish Kumar Nagireddy <Satish.nagireddy.nagireddy@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_disp.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_disp.c b/drivers/gpu/drm/xlnx/zynqmp_disp.c
index adb0909d917c..68d34dfc4d4e 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_disp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_disp.c
@@ -2656,6 +2656,9 @@ zynqmp_disp_plane_atomic_async_update(struct drm_plane *plane,
 	struct drm_plane_state *old_state =
 		drm_atomic_get_old_plane_state(new_state->state, plane);
 
+	if (plane->state->fb == new_state->fb)
+		return;
+
 	 /* Update the current state with new configurations */
 	swap(plane->state->fb, new_state->fb);
 	plane->state->crtc = new_state->crtc;
-- 
2.31.1

